### PR TITLE
fix: Nil pointer on build failure recovery

### LINKED ISSUE #804: Nil pointer on build failure recovery
I wanted to simulate a build failure so I did:

```
$ kamel run --dependency=camel-foobar examples/hello.xml
```

Which made the operator crash:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x12c9b56]
goroutine 412 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/james/go/src/github.com/apache/camel-k/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x108
panic(0x1490540, 0x245c110)
	/usr/share/go/src/runtime/panic.go:513 +0x1b9
github.com/apache/camel-k/pkg/controller/build.(*errorRecoveryAction).Handle(0xc0005b8500, 0x182ae60, 0xc000042028, 0xc000638480, 0xc00089dc58, 0x1, 0x1)
	/home/james/go/src/github.com/apache/camel-k/pkg/controller/build/recovery.go:100 +0x346
github.com/apache/camel-k/pkg/controller/build.(*ReconcileBuild).Reconcile(0xc000100a80, 0xc00079aca0, 0x9, 0xc000ad4f40, 0x18, 0x24714a0, 0x0, 0x0, 0x0)
	/home/james/go/src/github.com/apache/camel-k/pkg/controller/build/build_controller.go:188 +0x888
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000317900, 0x0)
	/home/james/go/src/github.com/apache/camel-k/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x18f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/home/james/go/src/github.com/apache/camel-k/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00034a0a0)
	/home/james/go/src/github.com/apache/camel-k/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00034a0a0, 0x3b9aca00, 0x0, 0x1, 0xc0005a03c0)
	/home/james/go/src/github.com/apache/camel-k/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbe
k8s.io/apimachinery/pkg/util/wait.Until(0xc00034a0a0, 0x3b9aca00, 0xc0005a03c0)
	/home/james/go/src/github.com/apache/camel-k/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/james/go/src/github.com/apache/camel-k/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x32a
```

[This](https://github.com/apache/camel-k/blob/master/pkg/controller/build/recovery.go#L100) seems to be the problem. `build.Status.Failure` wont have been initialized at this point.

### PR DESCRIPTION
**Pull request by jamesnetherton**
fixes #804

### CODE REVIEW

File: pkg/controller/build/recovery.go [hunk 1]
```diff
@@ -95,7 +95,9 @@ func (action *errorRecoveryAction) Handle(ctx context.Context, build *v1alpha1.B
 		return nil, nil
 	}
 
-	build.Status = v1alpha1.BuildStatus{}
+	build.Status = v1alpha1.BuildStatus{
+		Failure: build.Status.Failure.DeepCopy(),
+	}
```
[the remark this answers is not in the document]
Comment (jamesnetherton) [reply pkg/controller/build/recovery.go]: Yeah it is created beforehand. But build.Status is then reinitialized and currently it is not setting up the `Failure` field. Not sure whether [this](https://github.com/apache/camel-k/blob/master/pkg/controller/build/recovery.go#L98) is actually needed or not.

[the remark this answers is not in the document]
Comment (lburgazzoli) [reply pkg/controller/build/recovery.go]: yeah it does not make sense, seems a mistake

[the remark this answers is not in the document]
Comment (lburgazzoli) [reply pkg/controller/build/recovery.go]: the status update seems a mistake too

[the remark this answers is not in the document]
Comment (jamesnetherton) [reply pkg/controller/build/recovery.go]: Yep - that does seem wrong.

Ok - I'll fix those bits up and retest shortly.

File: pkg/controller/build/recovery.go [hunk 2]
```diff
@@ -87,8 +87,6 @@ func (action *errorRecoveryAction) Handle(ctx context.Context, build *v1alpha1.B
 		return nil, nil
 	}
 
-	build.Status = v1alpha1.BuildStatus{}
-	build.Status.Phase = v1alpha1.BuildPhaseInitialization
```
Comment (lburgazzoli) [inline pkg/controller/build/recovery.go]: I think resetting the phase is needed otherwise the build won'r be triggered again, isn't it ?

Comment (jamesnetherton) [reply pkg/controller/build/recovery.go]: Yep - Sorry, that wasn't meant to be removed. I misinterpreted your earlier comments. That's fixed now. 

However, when testing on top of the latest changes, if I provide an invalid dependency to `kamel run`, the build is actually somehow successful. That's definitely not the behaviour when I force HEAD back to cb988c7543ed9a17a44c53c35c50cee4a9ef19c5, the build fails and the recovery loop kicks in as expected.

Comment (lburgazzoli) [reply pkg/controller/build/recovery.go]: interesting, do you have time to investigate ?

Comment (jamesnetherton) [reply pkg/controller/build/recovery.go]: Yeah, I'll take a look tomorrow.

Comment (jamesnetherton) [reply pkg/controller/build/recovery.go]: I think we're good to merge this now that we have #807 fixed.

**Pull request merged**